src/Orleans.Runtime.Legacy/Configuration/LegacyClusterConfigurationExtensions.cs

```diff
@@ -231,20 +232,17 @@ private static void AddLegacyClusterConfigurationSupport(IServiceCollection serv
             });
 
             // add grain service configs as keyed services
-            short id = 0;
             foreach (IGrainServiceConfiguration grainServiceConfiguration in configuration.Globals.GrainServiceConfigurations.GrainServices.Values)
             {
-                services.AddSingletonKeyedService<long, IGrainServiceConfiguration>(id++, (sp, k) => grainServiceConfiguration);
+                var type = Type.GetType(grainServiceConfiguration.ServiceType);
+                services.AddSingletonKeyedService(type, (sp, k) => grainServiceConfiguration);
             }
+
             // populate grain service options
-            id = 0;
-            services.AddOptions<GrainServiceOptions>().Configure<GlobalConfiguration>((options, config) =>
+            foreach(IGrainServiceConfiguration grainServiceConfiguration in configuration.Globals.GrainServiceConfigurations.GrainServices.Values)
             {
-                foreach(IGrainServiceConfiguration grainServiceConfiguration in config.GrainServiceConfigurations.GrainServices.Values)
-                {
-                    options.GrainServices.Add(new KeyValuePair<string, short>(grainServiceConfiguration.ServiceType, id++));
-                }
-            });
+                services.AddSingleton<IGrainService>(sp => GrainServicesSiloBuilderExtensions.ConstructGrainService(Type.GetType(grainServiceConfiguration.ServiceType), sp));
```

Comment:
add a method ` public static IServiceCollection AddGrainService<T>(this IServiceCollection builder) ` in GrainServicesSiloBuilderExtensions.cs and use it here. And also `ISiloHostBuilder AddGrainService<T>(this ISiloHostBuilder builder) ` can reuse it. 

for increase code reuse

Reply:
Please no new methods on `IServiceCollection`, especially ones which don't apply to both silo and client.

A generic method wont work here, since we don't statically know the type `T` to use as a parameter.